sys/select.c

```diff
@@ -48,12 +48,26 @@ int select(int nfds, fd_set *rd, fd_set *wr, fd_set *ex, struct timeval *to)
 	size_t i, n;
 	int err;
 
+	if (nfds < 0 || nfds > FD_SETSIZE) {
+		errno = EINVAL;
+		return -1;
+	}
+
+	if ((to != NULL) && (to->tv_usec < 0 || to->tv_usec > 999999)) {
+		errno = EINVAL;
+		return -1;
+	}
+
 	for (n = i = 0; i < nfds; ++i)
 		if (SFD_ISSET(i, rd) || SFD_ISSET(i, wr) || SFD_ISSET(i, ex))
 			++n;
 
-	if (!n)
-		return usleep(poll_timeout(to));
+	timeout = (time_t)poll_timeout(to);
```

Comment:
`to == NULL` is accepted, but it will cause bad stuff below - `timeout = -1` then and `usleep(-1) `waits for few thousand years. It's not super bad right now, as `time_t` is unsigned, but it might become signed in the future and I guess negative time would mean no wait at all. I guess we don't want a timeout then, but I guess we should rather set `timeout` to zero if `to` is `NULL`

Reply:
Or even better, we should check retval for `< 0` before assigning the `timeout`

Reply:
Well, it depends whether we classify `timeout = -1` and `usleep(-1)`  as bad. I initially thought that current `usleep(-1)` is a bug, as using select just to emulate `usleep` by passing empty sets and some timeout is okay, but passing NULL timeout just to hang approx. forever seemed pointless to me. Then I checked how glibc handles `select(_, NULL, NULL, NULL, NULL)` and it hangs just like that. Actually, POSIX implies that a NULL timeout should cause select to block indefinitely waiting for any fd to become ready. It doesn't restrict the sets to be non-empty in this case, so the select is allowed to block forever, so the glibc behavior makes sense (although IMHO still seems a bit pointless)

Reply:
Yeah, I agree that behaviour is ok, but our sleep expects 0 for such case, not negative (in reality really big, as time_t is unsigned) value

Reply:
What I mean is something along the lines of:
```
int ret = poll_timeout(to);
timeout = ret < 0 ? 0 : (time_t)ret;
```

Reply:
This is not ideal still, because if tv_sec >= INT_MAX/1000 then we will sleep indefinitely. Not a really use-case though